interfaces/apparmor/apparmor.go

```diff
@@ -83,22 +85,49 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
 	return nil
 }
 
-// UnloadProfiles removes the named profiles from the running kernel.
-//
-// The operation is done with: apparmor_parser --remove $names...
-// The binary cache file is removed from /var/cache/apparmor
+// UnloadProfiles is meant to remove the named profiles from the running
+// kernel and then remove any cache files. Importantly, we can only unload
+// profiles when we are sure there are no lingering processes from the snap
+// (ie, forcibly stop all running processes from the snap). Otherwise, any
+// running processes will become unconfined. Since we don't have this guarantee
+// yet, leave the profiles loaded in the kernel but remove the cache files from
+// the system so the policy is gone on the next reboot.
 func unloadProfiles(names []string, cacheDir string) error {
 	if len(names) == 0 {
 		return nil
 	}
 
-	args := []string{"--remove"}
-	args = append(args, names...)
-	output, err := exec.Command("apparmor_parser", args...).CombinedOutput()
-	if err != nil {
-		return fmt.Errorf("cannot unload apparmor profile: %s\napparmor_parser output:\n%s", err, string(output))
+	/* TODO: uncomment when no lingering snap processes is guaranteed
+	// By the time this function is called, all the profiles (names) have
+	// been removed from dirs.SnapAppArmorDir, so to unload the profiles
+	// from the running kernel we must instead use sysfs and write the
+	// profile names one at a time to
+	// /sys/kernel/security/apparmor/.remove (with no trailing \n).
+	apparmorSysFsRemove := "/sys/kernel/security/apparmor/.remove"
+	if _, err := os.Stat(appArmorSysFsRemove); err != nil && os.IsNotExist(err) {
+	        return fmt.Errorf("cannot unload apparmor profile: %s does not exist\n", appArmorSysFsRemove)
 	}
-	if err := osutil.UnlinkMany(cacheDir, names); err != nil && !os.IsNotExist(err) {
+	for _, n := range names {
+	        // ignore errors since it is ok if the profile isn't removed
+	        // from the kernel
+	        ioutil.WriteFile(appArmorSysFsRemove, []byte(n), 0666)
+	}
+	*/
+
+	// AppArmor 2.13 and higher has a cache forest while 2.12 and lower has
+	// a flat directory (on 2.12 and earlier, .features and the snap
+	// profiles are in the top-level directory instead of a subdirectory).
+	// With 2.13+, snap profiles are not expected to be in every
+	// subdirectory, so don't error on ENOENT but otherwise if we get an
+	// error, something weird happened so stop processing.
+	if li, err := filepath.Glob(filepath.Join(cacheDir, "*/.features")); err == nil && len(li) > 0 { // 2.13+
```

Comment:
Hmm...

I'm not very happy about this. Is there any tool that would be compatible across apparmor versions that would reliably remove items from the cache?

Reply:
No. There is the libapparmor API (C library) but it only supports working with the current cache, not the historic caches. This is no different from previous behavior where didn't use some API and just managed the caches ourselves; it is just that there are now more cache directories (one per kernel abi). Note that I discussed this with apparmor upstream and the .features file is guaranteed to exist in each cache directory since it is required for the parser to use it. I keyed off this file instead of doing a simple glob on the filenames since I didn't want to interfere with other directories that might be in there (of which there shouldn't be any, but was being safe). FWIW, dh_apparmor is going to be modified to do the exact same thing here.